### ISSUE TITLE
[HOLD for payment 2024-03-18] [$250] Room - Flag as offensive function is absent in the conversation, but present in the thread

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#36045]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.38-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause internal team
**Slack conversation:**

## Action Performed:
1. Open https://staging.new.expensify.com/

2. Log in under your HT account A
3. In incognito mode, open https://staging.new.expensify.com/
4. Log in with a different account B
5. Under user A, create a WS
6. Invite user B to the created WS
7. Under user A, send a message to the @announce room
8. In the room settings, enable Who can post - Admins only
9. Under user B, right-click on a message from user A
10. Note that the Flag as offensive function is not available
11. Create a thread with this message
12. In the thread, right-click on the parent message
## Expected Result:
The Flag as offensive function must be present in the conversation and is present in the thread. Or absent in the conversation and thread, if so intended.
## Actual Result:
Flag as offensive function is absent in the conversation, but present in the thread

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/38435837/12faa2e6-f9ed-4165-aac4-a664f6de2c30

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~015264f638251d5c13</li>
        <li>Upwork Job ID: 1755273971773313024</li>
        <li>Last Price Increase: 2024-02-28</li>
<li>Automatic offers: </li>
<ul>
<li>getusha | Reviewer | 0</li>
<li>dukenv0307 | Contributor | 0</li>
</ul></li>
    </ul>
</details>

**Comment by m-natarajan** [comment]
We think that this bug might be related #vip-vsb
CC @quinthar

**Comment by FitseTLT** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
 Flag as offensive function is absent in the conversation, but present in the thread

### What is the root cause of that problem?
Because `canComment` is true for the new thread the flag offensive is displayed
### What changes do you think we should make in order to solve the problem?

We need to add condition here
https://github.com/Expensify/App/blob/f6ba751301d6a6f4e5faa092e615d49fb02d2746/src/pages/home/report/ContextMenu/ContextMenuActions.tsx#L456-L461
```
            !(reportAction?.childReportID && reportAction.childReportNotificationPreference === 'hidden'),

```
### What alternative solutions did you explore? (Optional)

**Comment by FitseTLT** [comment]
@dukenv0307 I am not finished with my propsal and when I am finished will notify

**Comment by FitseTLT** [comment]
[Updated](https://github.com/Expensify/App/issues/36045#issuecomment-1932502768)

**Comment by Tony-MK** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Room - Flag as an offensive function is absent in the conversation but present in the thread

### What is the root cause of that problem?

The [`isAllowedToComment`](https://github.com/Expensify/App/blob/198ca2c503fdd6044ca02f7ae3bfad5d3878a90f/src/libs/ReportUtils.ts#L3932) determines whether to flag an offensive comment.

https://github.com/Expensify/App/blob/198ca2c503fdd6044ca02f7ae3bfad5d3878a90f/src/libs/ReportUtils.ts#L3925-L3933

Furthermore, the [`isAllowedToComment`](https://github.com/Expensify/App/blob/198ca2c503fdd6044ca02f7ae3bfad5d3878a90f/src/libs/ReportUtils.ts#L3932) will return false because `report?.writeCapability` is present and the member is not a part of the admin.

https://github.com/Expensify/App/blob/1b060129b537a206fdf170239809ad8520ebffd4/src/libs/ReportUtils.ts#L1068-L1069

### What changes do you think we should make in order to solve the problem?


We can refuse to display the flag button if the policy only allows admins to comment.

Instead of the report, use the root workspace report when using the [`isAllowedToComment`](https://github.com/Expensify/App/blob/198ca2c503fdd6044ca02f7ae3bfad5d3878a90f/src/libs/ReportUtils.ts#L3932) in the [`canFlagReportAction`](https://github.com/Expensify/App/blob/198ca2c503fdd6044ca02f7ae3bfad5d3878a90f/src/libs/ReportUtils.ts#L3908) function to the function to perform when the chats theard which are children in the policy but the parent report is still a thread and not the root workspace report.

https://github.com/Expensify/App/blob/198ca2c503fdd6044ca02f7ae3bfad5d3878a90f/src/libs/ReportUtils.ts#L3932

Therefore, we should replace the [`isAllowedToComment`](https://github.com/Expensify/App/blob/198ca2c503fdd6044ca02f7ae3bfad5d3878a90f/src/libs/ReportUtils.ts#L3932) with the condition below.

```
isWorkspaceThread(report) ? isAllowedToComment(getRootParentReport(report)) :  isAllowedToComment(report)
```

### What alternative solutions did you explore? (Optional)

On the other hand, we could remove the [`isAllowedToComment`](https://github.com/Expensify/App/blob/198ca2c503fdd6044ca02f7ae3bfad5d3878a90f/src/libs/ReportUtils.ts#L3932) and allow everyone to flag the comment just like reacting, replying, and marking it unread.

Like the mentioned buttons, even if the user is unauthenticated,  the user will log in after pressing the flag comment button.

**Comment by getusha** [comment]
@Tony-MK's [proposal](https://github.com/Expensify/App/issues/36045#issuecomment-1936826098) looks good to me.
🎀 👀 🎀 C+ Reviewed.

**Comment by Tony-MK** [comment]
Could we get a clarification on the expected result?

Should we show the offensive Flag in the room thread or not?

Also, should we apply the fix to the threads inside the room thread?

**Comment by JmillsExpensify** [comment]
Hmm, yeah I think this is by design right? The admin-only post is not flagged by design, but the thread is a distinct convo from the parent, so it can be. @dangrous Do you mind confirming?

**Comment by dangrous** [comment]
Hi! So @JmillsExpensify and I discussed separately:

1. We're going to bring this to a broader discussion, we're not sure on the action here.
2. Yes, this should be able to be flagged.

**Comment by JmillsExpensify** [comment]
Ah apologies, I lost this in the shuffle. I will post a discussion now.

**Comment by JmillsExpensify** [comment]
Just posted, so I will update this issue once we align internally.

**Comment by JmillsExpensify** [comment]
Still waiting on alignment internally.

**Comment by youssef-lr** [comment]
Not overdue.

**Comment by youssef-lr** [comment]
No update yet.

**Comment by JmillsExpensify** [comment]
We landed on not being able to flag admin posts for now.

**Comment by youssef-lr** [comment]
@dukenv0307 has the correct proposal here. Adjusting price as this is low priority and a simple fix.

**Comment by Tony-MK** [comment]
@youssef-lr, I believe my [proposal](https://github.com/Expensify/App/issues/36045#issuecomment-1932535644) fixes the issue because the selected [proposal](https://github.com/Expensify/App/issues/36045#issuecomment-1936826098) only prevents non-admins not flagging the child thread but not deeper workspace chat threads.

This was the main reason why I posted my [proposal](https://github.com/Expensify/App/issues/36045#issuecomment-1932535644).

Furthermore, the [comments](https://github.com/Expensify/App/issues/36045#issuecomment-1937920034) about my proposal are false.

Could you re-evaluate my [proposal](https://github.com/Expensify/App/issues/36045#issuecomment-1932535644)?

**Comment by youssef-lr** [comment]
@Tony-MK This is your suggested change:

>Therefore, we should replace the [isAllowedToComment](https://github.com/Expensify/App/blob/198ca2c503fdd6044ca02f7ae3bfad5d3878a90f/src/libs/ReportUtils.ts#L3932) with the condition below.
>
> `isWorkspaceThread(report) ? isAllowedToComment(getRootParentReport(report)) :  isAllowedToComment(report)`

As @dukenv0307 mentioned, I think `isWorkspaceThread(report)` will return `true` for every reply in a thread from an an admins-only room. So `canFlagReportAction` will always return ` isAllowedToComment(getRootParentReport(report))` which should be `false` for non-admins in a thread, even though they're allowed to comment on the thread. Anything I'm missing here?

**Comment by Tony-MK** [comment]
After checking back on my proposal, I used the getRootParentReport function to get correct report to pass to isAllowToComment because the thread report can not be used to evaluate whether a user can flag.

That's why the parent admin report doesn't allow a user to flag a messge but can the message thread.

Therefore, it should return true for non admin rooms.

I tested it and worked but could I not be grasping something?

**Comment by JmillsExpensify** [comment]
@getusha Do you mind filling out the BZ checklist above? In the meantime, I believe the payment summary is:

 - Contributor: @dukenv0307 $500
 - Contributor+: @getusha $500

**Comment by getusha** [comment]
>Contributor: @dukenv0307 $500
Contributor+: @getusha $500

@JmillsExpensify it is $250 based on https://github.com/Expensify/App/issues/36045#issuecomment-1968171833

**Comment by JmillsExpensify** [comment]
Thanks. I updated the payment summary above and paid all contracts in Upwork.
